### ISSUE TITLE
Component metadata rules should allow adding/removing dependencies

### ISSUE THREAD

**Issue by jjohannes** [gradle/gradle#3157]
## Use Cases

There are two main use cases:

1. The published metadata is _wrong_, where wrong can mean different things. For instance:
-A dependency to an implementation was added for _convenience_ in the `runtime` configuration, although a dependency to the API would be sufficient (e.g. `slf4j-simple` vs. `slf4j-api`).
-A version range is too strict. For example, when the dependency was defined with a closed range at the time of publishing, but newer versions of the dependency also work.
-The author of the dependency just made a mistake but does not (for whatsoever reason) release a new version that fixes the published metadata. 
2. Variants should be defined for a published module where the variant information is not in the published metadata. It might be necessary to assign only subsets of the dependencies to different variants. While the ability to define variants will be tackled in #3159, the ability to modify dependencies is a prerequisite. 

## Spec

- **Dependencies of an external module can be accessed for inspection**
  - In order to decide how to manipulate dependencies, it might be necessary to inspect what is there.
- **A dependency can be replaced with one or more other dependencies**
  - I can be necessary to replace one dependency with another. For example a patch or a fork published under different coordinates or a different implementation of the same specification (e.g. a different logging implementation).
  - API wise, it is ok to provide `remove()` and `add()` functionalities for single dependencies that can be combined by the user to achieve the desired replacement.
- **A dependency can be removed**
  - Part of the _change dependency_ story. 
  - Also useful for special cases where metadata is _wrong_.
-  **A dependency can be added**
   - Part of the _change dependency_ story. 
   - Also useful for special cases where metadata is _wrong_.

This does **NOT** cover (follow up work)
- Modify dependencies of selected configurations or variants only.

## Proposed API
```
dependencies {
    c1 'org.gradle.test.g1:a1:2.0'

    components {
        withModule("org.gradle.test.g1:a1") {
            withVariant("default") {
                withDependencies { dependencies ->
                    dependencies.removeAll { it.version == "1.0" }
                    dependencies.add("org.gradle.test.g2:a2-patch:2.0") {
                        optional = true
                    }
                }
            }
        }
    }
}
```

## Implementation
See: https://github.com/gradle/gradle/pull/3158

Extending the component metadata rules to modify dependencies seems straightforward. These rules are evaluated when a new module is added to the dependency graph, but before we try to resolve the dependencies of that module. So no adjustment is needed in the resolution.
In essence, we can make the list of dependencies in `MutableModuleComponentResolveMetadata` and its subclasses mutable and then add access and modification methods (see API proposal above) to `ComponentMetadataDetails`, which is implemented as an adapter on top of `MutableModuleComponentResolveMetadata`.

## Performance Impact

The rules are evaluated in every run of a build that resolves modules which are matched by the rules. The main performance impact is expected by the creation of additional objects that represent the dependencies. In particular rules that do not manipulate dependencies, and which already exist today, should not be impacted. Therefore, all dependency access and manipulation will be implemented in a lazy way. I.e. objects that present the dependency metadata will only be created if the dependency metadata is accessed in the rule.
In general, when the rules become more powerful and are used more extensively, they will get more expensive in general. E.g. additional metadata information could be queried from another source and added in the rule. These general thoughts are out of scope of this issue and will be further evaluated in #3160.

**Comment by oehme** [comment]
I added two comments to the spike.I'm wondering why we would have special `remove` and `update` methods. Why would we not allow a more direct interaction?

```
dependencies.removeAll { it.name == "guava" }
```

or

```
dependencies.findAll { it.name = "guava"}.each {
  it.version = '21.0'
}
```

**Comment by jjohannes** [comment]
What I proposed was the simpler solution implementation wise. We don't have a a public interface for (mutable) dependency metadata. We would have to introduce that. Which we can of course.

And since we need some kind of `add` method, having `remove` and `update` also felt like one possible way to go.

But I agree that there are some advantages in the "more direct interaction". What about `add` then? Should we have a special type of collection such that we can do:

```
dependencies.add("com.google.guava:guava:1.0") {
   //more details
}
```

**Comment by jjohannes** [comment]
Summary of discussing this further with @oehme:
- Scope of this issue: add/remove (NOT update)
- Implementaion
  - We need to make sure we copy the dependency metadata and do not modify the original
  - Introduce a new public interface for inspecting/creating component metadata
    - Introduce a `ComponentDependenciesMetadataDetailsAdapter` and `ComponentDependencyMetadataAdapter` along the lines of `ComponentMetadata(Details)Adapter`. They implement the new public interfaces.
     - `ComponentDependencyMetadata` allows to access all known properties of the metadata. (For the `add` method below we also need a mutable ComponentDependencyMetadataDetails interface.)
     - `ComponentDependenciesMetadata` extends `List<ComponentDependencyMetadataDetails>` and adds an `add(dependencyNotation, Action<MutableComponentDependencyMetadata> configureAction)` method.
   - Adapters have to be created lazily

**Comment by oehme** [comment]
Can you update the expected DSL above, please?

**Comment by jjohannes** [comment]
I updated the API (looks good so far) and the performance section in the issue description.

**Comment by jjohannes** [comment]
I update #3158 following what we discussed. Feedback welcome @oehme and @bigdaz.

**Comment by adammurdoch** [comment]
Is there a design doc for this? Or at the very least a list of the use cases that we're going to tackle?

**Comment by jjohannes** [comment]
@adammurdoch I added _use cases_ to the description of this issue.

**Comment by adammurdoch** [comment]
There's not really any such thing as "the dependencies" of a component. The dependencies that are used are always a function of the context.

I think the API we offer here should already start to address this in some way. For example, by providing details about the set of dependencies that are about to be used to traverse the graph, rather than just a "getDependencies()" method. I'm not 100% convinced trying to reuse component metadata rules as they are now is the right place to wire in the kind of logic we're aiming for.

I would also try hard to make the API to manipulate the dependency set in a metadata rule as consistent as possible with the API to manipulate the dependency set in the DSL.

**Comment by adammurdoch** [comment]
I would probably use 2 kinds of rules:

1. Rules that allow us to determine whether we should select the component, which would include details about the variants available for the component and other information about the component as a whole, but nothing about the dependencies or files. The existing metadata rules and metadata suppliers, and maybe selection and replacement rules too, would be merged into this bucket.
2. Rules that allow us to use a particular variant just prior to traversing it, which would include information about the dependencies and files of that variant. This would be a new bucket.

The separation allows expensive work to finalise the details of a variant to be deferred until required, including fetching/calculating the inputs to this work.

A variation on this theme would be to allow the rules in bucket 1 to express information about the variants when the rule can do this cheaply. Then, we might just start with bucket 1, and rules will have to be careful to adjust the dependencies and files for _all_ variants without knowing which of these will be required, if any.

In general, I think the idea of trying to collapse all of the logic into one kind of rule is too coarse grained to work well. I would go for specialised rules that target one particular question (and make the pattern consistent for this too).

**Comment by jjohannes** [comment]
@adammurdoch I had a longer discussion with @oehme.

We agree that we should manipulate the dependencies in its _context_ (i.e. configuration or variant). We were discussing if we should start with giving access to configurations, which would be in line with the `dependencies {}` API as it is at the moment. E.g:

```
   components {
        withModule("org.gradle.test.g1:a1") {
            configuration ("compile") {
                dependencies.removeAll { it.version == "1.0" }
                dependencies.add("org.gradle.test.g2:a2-patch:2.0") {
                   force = true
                }
            }
        }
    }
```
 Or if we should go with variants directly. E.g.:
```
   components {
        withModule("org.gradle.test.g1:a1") {
            outgoing.variants.default { // instead of selecting by name we could also match attributes, 
                                        // and provide information about Ivy configs and Maven scopes as attributes
                dependencies.removeAll { it.version == "1.0" }
                dependencies.add("org.gradle.test.g2:a2-patch:2.0") {
                   force = true
                }
            }
        }
    }
```
We would then need to make sure that we automatically provide a default variant for each configuration.

I currently think we should go with `variants` directly, as you latest additions already introduced them as part of the new metadata format. 
Do you think this is going in the right direction?

**Comment by bigdaz** [comment]
We should _not_ be referencing `configurations` within component metadata rules. The second option looks like a step in the right direction to me.

**Comment by adammurdoch** [comment]
i agree, let's try not to expose configurations if we can avoid it.

**Comment by jjohannes** [comment]
Getting back to the idea of having different types of rules that apply at different points in the dependency resolution process mentioned in this comment above https://github.com/gradle/gradle/issues/3157#issuecomment-337614699

> 2. Rules that allow us to finalize a particular variant just prior to traversing it, which would include information about the dependencies and files of that variant. This would be a new bucket.

I think the use cases in this issue would require this kind of rules. And we could consider applying them later. For instance right before the dependencies of a node in the graph are analysed here:
https://github.com/gradle/gradle/blob/f14be56910762f3c3c551e2e26d9e3451d454a64/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java#L172-L186

We could do a `applyRules()` right before the dependencies are iterated, which takes a different kind of rule, lets call them _Variant Metadata Rules_ for now, with `RuleAction<VariantMetadataDetails>`.
Where `VariantMetadataDetails` then gives access to dependencies. Which can be implemented as in the PR associated to the ticket. Basically instead of adding `getDependencies()` to `ComponentMetadataDetails`, we introduce `VariantMetadataDetails.getDependency()`, and move the interpretation of the rules to a later (the latest possible) moment.

Internally, the implementation currently works with configurations instead of variants (`ConfigurationMetadata`). But we can assume a 1-to-1 mapping for now. Eventually we can pull up the notion of variants from `VariantAwareConfigurationMetadata` and assume one default varian per configurationt for metadata where variants can not be specified.
If at one point we support different dependencies per variant, the code linked above would be one of the places we need to touch in the resolution engine to create a new `EdgeState` per variant rather than per configuration.

If we start here, we could do a _clean_ implementation of mutable ConfigurationMetadata and VariantMetadata here. And possibly reuse that later to improve the implementation around Component Metadata Rules (#3241) - which would be the other type of rule.

**Comment by oehme** [comment]
The use case doesn't strictly require this kind of rule, though it may be more efficient to split it that way in the future. 

I think a good start for now would be to replace `MutableComponentResolveMetadata#getDependencies()` with `MutableComponentResolveMetadata#getVariants().getDependencies()`. We can adjust the clients of the former to use the latter. We can then have different implementations for `getVariants()` for Ivy and Maven and store them in different ways internally, but still offer one consistent API to the user, which allows them to modify dependencies of a variant and later adding/removing/modifying variants themselves.

**Comment by jjohannes** [comment]
Eventually we decided to go into the direction of "multiple kind of rules" and offer the following interface:

```
withModule("org.gradle.test.g1:a1") {
    withVariant("default") {
        withDependencies { dependencies ->
            dependencies.removeAll { it.version == "1.0" }
            dependencies.add("org.gradle.test.g2:a2-patch:2.0") {
                optional = true
            }
        }
    }
}
```

`withVariant {}` and `withDependencies {}` create separate actions that will be executed lazily or not at all if not needed. `withDependencies {}` (inside a `withVariant {}`) is then only executed if the variant is selected and its dependencies are queried.

We also discussed adding a `withAllVariants {}` but I am not fully convinced that we want that. Eventually, adjusting existing variants should only be done if there is something fundamentally wrong with the metadata of one variant (i.e. wrong in most contexts). So adjustments should be done very selective and not just "for everything".
In most cases, we should rather add new variants for using a published component in a new context. (Which will be supported in a follow up issue.)
